### PR TITLE
add consistency tests to ModelChain.dc_model

### PR DESCRIPTION
**Pull request by cwhanse**
pvlib python pull request guidelines
====================================

Thank you for your contribution to pvlib python! You may delete all of these instructions except for the list below.

You may submit a pull request with your code at any stage of completion.

The following items must be addressed before the code can be merged. Please don't hesitate to ask for help if you're unsure of how to accomplish any of the items below:

 - [x] Closes issue #417 
 - [x] I am familiar with the [contributing guidelines](http://pvlib-python.readthedocs.io/en/latest/contributing.html).
 - [x] Fully tested. Added and/or modified tests to ensure correct behavior for all reasonable inputs. Tests (usually) must pass on the TravisCI and Appveyor testing services.
 - [x] Updates entries to `docs/sphinx/source/api.rst` for API changes.
 - [x] Adds description and name entries in the appropriate `docs/sphinx/source/whatsnew` file for all changes.
 - [x] Code quality and style is sufficient. Passes ``git diff upstream/master -u -- "*.py" | flake8 --diff``
 - [x] New code is fully documented. Includes sphinx/numpydoc compliant docstrings and comments in the code where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.

Brief description of the problem and proposed solution (if not already fully described in the issue linked to above):

Tests for consistency between `module` parameters and `dc_model` choice.

### CODE REVIEW

File: pvlib/modelchain.py [hunk 1]
```diff
@@ -365,16 +365,56 @@ def dc_model(self, model):
...
+                                     'or more required parameters: ' + \
+                                     str(missing_params))
+                else:
+                    self._dc_model = self.pvwatts_dc
             else:
                 raise ValueError(model + ' is not a valid DC power model')
         else:
             self._dc_model = partial(model, self)
 
+    def _sapm_param_set():
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/modelchain.py]: At what point in `modelchain.py` should this function be imported?

File: pvlib/modelchain.py [hunk 2]
```diff
@@ -365,16 +365,56 @@ def dc_model(self, model):
...
+        return set(['A0', 'A1', 'A2', 'A3', 'A4', 'B0', 'B1', 'B2', 'B3', 'B4', 
+                    'B5', 'C0', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6', 'C7', 'Isco',
+                    'Impo', 'Aisc', 'Aimp', 'Bvoco', 'Mbvoc', 'Bvmpo', 'Mbvmp',
+                    'N', 'Cells_in_Series', 'IX0', 'IXX0', 'FD'])
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/modelchain.py]: yes

File: pvlib/modelchain.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wholmgren) [reply pvlib/modelchain.py]: Probably at the top with the rest of the imports.

**Comment by cwhanse** [comment]
I'm considering the following change, in which `infer_dc_model` is always followed by the logic in `dc_model` to validate the parameters.  To me this is better than having `infer_dc_model` do its own consistency check on the full parameter list.

`    def dc_model(self, model):

       # guess at model if None
        if model is None:
            self._dc_model, model = self.infer_dc_model()
        # Set model and validate parameters
        if isinstance(model, str):
            model = model.lower()
            # validate module parameters
            missing_params = DC_MODEL_PARAMS[model] - \
                                    set(self.system.module_parameters.keys())
            if missing_params: # some parameters are not in module.keys()
                raise ValueError(model + ' selected for the DC model but '
                                     'one or more required parameters '
                                     ' are missing : ' +
                                     str(missing_params))
            if model == 'sapm':
                self._dc_model = self.sapm
            elif model == 'singlediode':
                self._dc_model = self.singlediode
            elif model == 'pvwatts':
                self._dc_model = self.pvwatts_dc
            else:
                raise ValueError(model + ' is not a valid DC power model')
        else:
            self._dc_model = partial(model, self)

    def infer_dc_model(self):
        # returns both model function object and model string, could drop
        # model function object in the future since the model function object
        # will be set in dc_model after validating parameter consistency
        params = set(self.system.module_parameters.keys())
        if set(['A0', 'A1', 'C7']) <= params:
            return self.sapm, 'sapm'
        elif set(['a_ref', 'I_L_ref', 'I_o_ref', 'R_sh_ref', 'R_s']) <= params:
            return self.singlediode, 'singlediode'
        elif set(['pdc0', 'gamma_pdc']) <= params:
            return self.pvwatts_dc, 'pvwatts_dc'
        else:
            raise ValueError('could not infer DC model from '
                             'system.module_parameters')

`

**Comment by wholmgren** [comment]
@cwhanse that's ok with me. Your comment says "could drop model function object in the future" -- why not now?

**Comment by cwhanse** [comment]
It’s an API change for the method, I’m hesitant to possibly break someone’s code without a warning although the risk here is probably zero.

**Comment by wholmgren** [comment]
It's an API change either way, but I agree the risk is near 0 and not worth a deprecation cycle.

**Comment by adriesse** [comment]
Just picking up on the last few comments out of context...

I think it would be nice to users to generate warnings related to API changes, provided you can detect that they're trying to use something form the old API, _**and**_ to keep those warnings around for a good long time.  This will be especially useful for occasional users, who may not update everything all the time, and new users, who may pick up examples based on any past version, and old users who can't remember stuff.

**Comment by wholmgren** [comment]
@adriesse 0.6.0 will print deprecation warnings for the vast majority of API changes. I'd like to keep the warnings around "long enough" but I have no idea what that or "a good long time" means in practice! Realistically, we're looking at 6-12 months for the current pace of 0.x releases.

For the specific API change discussed here, I'd rather not go through a deprecation cycle because I don't think anyone is explicitly using this method so It's not worth the effort.

**Comment by cwhanse** [comment]
Ready to merge, I think. #463 and #487 need to build on top of these changes.

File: pvlib/test/test_modelchain.py [hunk 3]
```diff
@@ -376,7 +408,7 @@ def test_invalid_models(model, system, location):
     kwargs[model] = 'invalid'
     with pytest.raises(ValueError):
         mc = ModelChain(system, location, **kwargs)
-
+
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/test/test_modelchain.py]: Done

File: pvlib/modelchain.py [hunk 4]
```diff
@@ -360,29 +361,44 @@ def dc_model(self):
...
-            elif model == 'pvwatts':
-                self._dc_model = self.pvwatts_dc
+            if model in DC_MODEL_PARAMS.keys():
+                # validate module parameters
+                missing_params = DC_MODEL_PARAMS[model] - \
+                                        set(self.system.module_parameters.keys())
+                if missing_params: # some parameters are not in module.keys()
+                    raise ValueError(model + ' selected for the DC model but '
+                                         'one or more required parameters '
+                                         ' are missing : ' +
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply pvlib/modelchain.py]: Removed

File: docs/sphinx/source/whatsnew/v0.6.0.rst [hunk 5]
```diff
@@ -101,6 +101,7 @@ Enhancements
 * Add irradiance.gti_dirint function. (:issue:`396`)
 * Add irradiance.clearness_index function. (:issue:`396`)
 * Add irradiance.clearness_index_zenith_independent function. (:issue:`396`)
+* Add checking for consistency between module_parameters and dc_model. (:issue:`417`)
```
[the remark this answers is not in the document]
Comment (cwhanse) [reply docs/sphinx/source/whatsnew/v0.6.0.rst]: Added to `whatsnew` for `v0.6.0.rst`

**Pull request merged**